src/protocol_edge.c

```diff
@@ -32,18 +32,54 @@
 #include "protocol.h"
 #include "utils.h"
 #include "xalloc.h"
+#include "submesh.h"
 
 extern bool node_write_devclass(meshlink_handle_t *mesh, node_t *n);
+extern bool node_write_submesh(meshlink_handle_t *mesh, node_t *n);
 
 bool send_add_edge(meshlink_handle_t *mesh, connection_t *c, const edge_t *e, int contradictions) {
 	bool x;
 	char *address, *port;
+	char *from_submesh, *to_submesh;
+	submesh_t *s = NULL;
+
+	if(c->node && c->node->submesh) {
+		if(e->from->submesh && (e->from->submesh != c->node->submesh)) {
+			return true;
+		}
+
+		if(e->to->submesh && (e->to->submesh != c->node->submesh)) {
+			return true;
+		}
+	}
+
+	if(e->from->submesh && e->to->submesh && (e->from->submesh != e->to->submesh)) {
+		return true;
+	}
 
 	sockaddr2str(&e->address, &address, &port);
 
-	x = send_request(mesh, c, "%d %x %s %d %s %s %s %d %x %d %d", ADD_EDGE, rand(),
-	                 e->from->name, e->from->devclass, e->to->name, address, port, e->to->devclass,
-	                 e->options, e->weight, contradictions);
+	if(e->from->submesh) {
+		from_submesh = e->from->submesh->name;
+	} else {
+		from_submesh = "0";
```

Comment:
I'd use `"."` instead of `"0"`, to match what we discussed earlier about the core mesh being like the path `.`, and the parent mesh being like `..`. Also, you need to ensure that you can never create a new submesh with one of these reserved names. You are already using `check_id()` on sub-mesh names when parsing ADD/DEL_EDGE messages. You should also call `check_id()` in `create_submesh()` to prevent the application from creating a sub-mesh with an invalid name. Furthermore, it might be good to add `#define CORE_MESH "."` to `submesh.h`, and use that instead of literal strings here.

Reply:
I declared a macro as #define CORE_MESH "." and iam using this macro